### PR TITLE
Convert single column binary predictions to two

### LINKED ISSUE #374: predict_type="prob" does not work with out_features=1
Hi @sebffischer I expected that I should be able to use measures to evaluate predictions, for binary classification with a torch learner with out_features=1. Here is an example adapted from #373 (thanks!)
```r
library(mlr3torch)
nn_bce_loss3 = nn_module(c("nn_bce_with_logits_loss3", "nn_loss"),
  initialize = function(weight = NULL, reduction = "mean", pos_weight = NULL) {
    self$loss = nn_bce_with_logits_loss(weight, reduction, pos_weight)
  },
  forward = function(input, target) {
    self$loss(input$reshape(-1), target$to(dtype = torch_float())-1)
  }
)
loss = nn_bce_loss3()
loss(torch_randn(10, 1), torch_randint(0, 1, 10))
task = tsk("sonar")
graph = po("torch_ingress_num") %>>%
  nn("linear", out_features = 1) %>>%
  po("torch_loss", loss = nn_bce_loss3) %>>%
  po("torch_optimizer") %>>%
  po("torch_model_classif",
     epochs = 1,
     batch_size = 32,
     predict_type="prob")
glrn = as_learner(graph)
glrn$train(task)
glrn$predict(task)
```
The code above has predict_type="prob" and out_features=1 so I am getting the following error on current main
```r
if(FALSE){#broke
  remotes::install_github("mlr-org/mlr3torch@6e99e02908788275622a7b723d211f357081699a")
  glrn$predict(task)
  ## Erreur dans dimnames(x) <- dn : 
  ##   la longueur de 'dimnames' [2] n'est pas égale à l'étendue du tableau
  ## This happened PipeOp torch_model_classif's $predict()
}
```
The error happens because the torch model outputs only one column, but some later code assumes there are two.

I hacked a solution that fixes this (see below), and I will file a PR.
```r
if(FALSE){#fix
  remotes::install_github("tdhock/mlr3torch@69d4adda7a71c05403d561bf3bb1ffb279978d0d")
  glrn$predict(task)
  ## <PredictionClassif> for 208 observations:
  ##  row_ids truth response
  ##        1     R        M
  ##        2     R        M
  ##        3     R        M
  ##      ---   ---      ---
  ##      206     M        M
  ##      207     M        M
  ##      208     M        M
}
```

### PR DESCRIPTION
**Pull request by tdhock**
Closes #374

### CODE REVIEW

**Comment by tdhock** [comment]
probably should add a test case

File: R/learner_torch_methods.R [hunk 1]
```diff
@@ -286,6 +286,13 @@ encode_prediction_default = function(predict_tensor, predict_type, task) {
   response = prob = NULL
   if (task$task_type == "classif") {
     if (predict_type == "prob") {
+      if (dim(predict_tensor)[2] == 1) {
+        N_data <- length(predict_tensor)
+        vec_dim <- c(N_data, 1)
+        pos_scores <- predict_tensor$reshape(vec_dim)
+        neg_scores <- torch::torch_zeros(N_data)$reshape(vec_dim)
```
Comment (tdhock) [inline R/learner_torch_methods.R:293]: for justification of using zero for negative scores before putting them through softmax, see my slides, section "Multi-class classification" https://github.com/tdhock/2023-08-deep-learning/blob/main/slides/torch-part1/06-backprop.pdf
image

**Comment by sebffischer** [comment]
I need to think a bit about this to be sure how to handle this nicely (e.g. via `nn("head")` handling everything as suggested in our other discussion). Thanks a lot for stress-testing the package and uncovering some shortcomings! :)

**Comment by sebffischer** [comment]
Also when we introduce this, I think it should be done consistently across classification learners.
Currently also the pretrained learners output a tensor of shape `(batch, n_classes)`, even when `n_classes == 2`.

The `loss` of a `LearnerTorch` (such as `lrn("classif.mlp", loss = t_loss("cross-entropy")` does not yet know which type of task it will be applied to during construction. I would like to avoid having an extra loss for binary cross-entropy and multi-class cross-entropy, as this is a bit annoying for benchmark experiments where a classification learner is sometimes applied to binary classification problems and other times to multi-class classification problems such as in:


```r
library(mlr3torch)
design = benchmark_grid( 
  # sonar is binary, iris multi-class
  tasks = tsks(c("sonar", "iris")),
  learners = lrn("classif.mlp", epochs = 10, batch_size = 16, loss = t_loss("cross_entropy")),
  resampling = rsmp("cv")
)
benchmark(design)
```

I think the solution is to make `t_loss("cross_entropy")` generate a `nn_bce_loss_with_logits` when it encounters a binary classification problem and otherwise a `nn_cross_entropy_loss`.
Also, maybe the information on how to load the targets should be attached to the loss function, which currently is not the case.

**Comment by tdhock** [comment]
agree about "make t_loss("cross_entropy") generate a nn_bce_loss_with_logits when it encounters a binary classification problem and otherwise a nn_cross_entropy_loss."

not sure I understand "load the targets should be attached to the loss function, which currently is not the case. Maybe even the prediction encoder should be attached to the loss"

**Comment by sebffischer** [comment]
> not sure I understand "load the targets should be attached to the loss function, which currently is not the case. Maybe even the prediction encoder should be attached to the loss"

When `mlr3torch` trains the model, the factor column from the task gets converted into a tensor. *how* this is done depends on the loss function as we have seen in this example. Because of this relationship I was thinking maybe the loss function should dictate how to load the target (`0-1` or `1-2`, int vs float).

The prediction encoder defines how a torch prediction tensor as output by the underlying network is converted to an `mlr3::Prediction` object. I am wondering from where we should get the information how to do this. I think a runtime check on the dimension as done here is a little hacky so ideally I would like to do this differently (and more generically).

**Comment by tdhock** [comment]
agree with "a runtime check on the dimension as done here is a little hacky so ideally I would like to do this differently (and solve it more generically)."

**Comment by sebffischer** [comment]
Superseded by https://github.com/mlr-org/mlr3torch/pull/385
